### PR TITLE
Check reference equality for strongly-typed equals in records

### LINKED ISSUE #51136: Update strongly-typed record Equals method
We need to add a short-circuit in the strongly-typed `Equals` method we generate for record equality to ensure that, if the same instance is being compared for equality, we immediately return true. See https://github.com/dotnet/csharplang/issues/39#issuecomment-678097433 for a benchmark of the perf consequences of this. /cc @jcouv.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #51136

Not sure if it's that simple.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: >F.ObjectEqual(F.This(), other) [](start = 38, length = 30)

Is this check duplicated in the equality operator now? Should it be removed from there?

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: @AlekseyTs The issue is that users can re-define `Equals`.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: > The issue is that users can re-define Equals.

And? The redefined Equals should still do the right thing.

---
In reply to: [574776221](https://github.com/dotnet/roslyn/pull/51149#discussion_r574776221) [](ancestors = 574776221)

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: @AlekseyTs Just tried to remove the check from `==` implementation, and now null comparison fails unfortunately :confused:

The *current* implementation looks like:

```csharp
        if ((object)r1 != r2) // if this is removed, (MyRecord)null == (MyRecord)null be false.
        {
            if ((object)r1 != null)
            {
                return r1.Equals(r2);
            }
            return false;
        }
        return true;
```

This was caught by the following test:

```csharp
        [Theory]
        [CombinatorialData]
        public void EqualityOperators_09(bool useImageReference)
        {
            var source1 =
@"
public record A(int X) 
{
}
";
            var comp1 = CreateCompilation(source1);

            var source2 =
@"
class Program
{
    static void Main()
    {
        Test(null, null);
        Test(null, new A(0));
        Test(new A(1), new A(1));
        Test(new A(2), new A(3));
    }

    static void Test(A a1, A a2)
    {
        System.Console.WriteLine(""{0} {1} {2} {3}"", a1 == a2, a2 == a1, a1 != a2, a2 != a1);
    }
}
";
            CompileAndVerify(source2, references: new[] { useImageReference ? comp1.EmitToImageReference() : comp1.ToMetadataReference() }, expectedOutput:
@"
True True False False
False False True True
True True False False
False False True True
").VerifyDiagnostics();
        }
```

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: > Just tried to remove the check from == implementation, and now null comparison fails unfortunately

I guess we could change the code gen to the following:
```
         if ((object)r1 != null)
            {
                return r1.Equals(r2);
            }

        return (object)r2 == null;
```

However, I am not sure if this is going to give us any noticeable improvement. 

---
In reply to: [575128469](https://github.com/dotnet/roslyn/pull/51149#discussion_r575128469) [](ancestors = 575128469)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: @dotnet/roslyn-compiler Thoughts? 

---
In reply to: [575311021](https://github.com/dotnet/roslyn/pull/51149#discussion_r575311021) [](ancestors = 575311021,575128469)

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: >F.LogicalOr [](start = 26, length = 11)

I think we should do similar change for the other consumer of the ```GenerateFieldEquals``` helper, the code that generates Equals for Anonymous Types. I also think that we should check/adjust code gen for VB Anonymous Types as well.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: @AlekseyTs That sounds reasonable. Should I start doing this and update tests? or do you want to wait for another team member to confirm?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: >> I think we should do similar change for the other consumer of the ```GenerateFieldEquals``` helper, the code that generates Equals for Anonymous Types. I also think that we should check/adjust code gen for VB Anonymous Types as well.

> That sounds reasonable. Should I start doing this and update tests? or do you want to wait for another team member to confirm?

I don't think we need to wait for the confirmation from other team members. If I was working on this change, I would just do this (the improvement of code gen for Anonymous Types). 

---
In reply to: [575327671](https://github.com/dotnet/roslyn/pull/51149#discussion_r575327671) [](ancestors = 575327671)

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: @AlekseyTs I made the change for anonymous types for C# and updated the tests.

Could you point me to where the parallel change for VB should go?

I looked at https://github.com/Youssef1313/roslyn/blob/records-ref-equality/src/Compilers/VisualBasic/Portable/Symbols/AnonymousTypes/SynthesizedSymbols/AnonymousType_EqualsMethodSymbol.vb but didn't find anything relevant. So I'm probably in the wrong place.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: > Could you point me to where the parallel change for VB should go?

See ```AnonymousType_IEquatable_EqualsMethodSymbol.GetBoundMethodBody``` in src\Compilers\VisualBasic\Portable\Binding\SyntheticBoundTrees\AnonymousTypeSyntheticMethods.vb.

---
In reply to: [575376850](https://github.com/dotnet/roslyn/pull/51149#discussion_r575376850) [](ancestors = 575376850)

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:158]: > Is this check duplicated in the equality operator now? Should it be removed from there?

In the email thread, I believe we agreed to duplicate the equality check.

**Comment by Youssef1313** [comment]
I'm having seemingly unrelated test failures locally (haven't yet checked if these happen on master or only this branch), but these doesn't happen in CI:

image

Is that a bug?

**Comment by AlekseyTs** [comment]
> Is that a bug?

This could be an issue with the tests, but more investigation is necessary to confirm. It appears that a sorted set is ordered differently for you locally. If sorting is indeed affected by some settings in your local environment, the tests should be adjusted to make sure this isn't happening and that the sorting is environment independent.

---
In reply to: [778409221](https://github.com/dotnet/roslyn/pull/51149#issuecomment-778409221) [](ancestors = 778409221)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

**Comment by jcouv** [comment]
>                 //  Method body:

it'd be great to update the snippet to reflect the new codegen.

---
Refers to: src/Compilers/CSharp/Portable/Compiler/AnonymousTypeMethodBodySynthesizer.cs:101 in da031eb. [](commit_id = da031eb585e3a4de7dd3b9eea0b41a5d2a11706a, deletion_comment = False)

**Comment by jcouv** [comment]
>                     // There are no base record types.

Same comment here and below (line 112)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordEquals.cs:81 in da031eb. [](commit_id = da031eb585e3a4de7dd3b9eea0b41a5d2a11706a, deletion_comment = False)

File: src/Compilers/VisualBasic/Portable/Binding/SyntheticBoundTrees/AnonymousTypeSyntheticMethods.vb
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Binding/SyntheticBoundTrees/AnonymousTypeSyntheticMethods.vb:216]: Possibly here too

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 5) with ask to keep comments in sync with code

**Comment by jcouv** [comment]
Merged/squashed. Thanks @Youssef1313!

**Pull request merged**
